### PR TITLE
Li thresholding port from imagej

### PR DESCRIPTION
**Pull request by jmetz**
Ported imagej version of Li thresholding (see refs in docstring).

### CODE REVIEW

**Comment by jmetz** [comment]
**Note:** Travis build is failing in unrelated areas, e.g. bottom of https://travis-ci.org/scikit-image/scikit-image/jobs/51085857 shows

    loading pickled environment... not yet created
    loading intersphinx inventory from http://scikit-learn.org/stable/objects.inv...
    Warning, treated as error:
    WARNING: intersphinx inventory 'http://scikit-learn.org/stable/objects.inv' not fetchable due to <class 'socket.error'>: [Errno 104] Connection reset by peer
    make[1]: *** [html] Error 1
    make[1]: Leaving directory `/home/travis/build/scikit-image/scikit-image/doc'
    make: *** [html] Error 2
    The command "tools/travis_script.sh" exited with 2.
    Done. Your build exited with 1.

File: skimage/filters/thresholding.py [hunk 1]
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+
+    Parameters
+    ----------
+    image : array
+        Input image.
+
+    Returns
+    -------
+    threshold : float
+        Upper threshold value. All pixels intensities that less or equal of
```
Comment (blink1073) [inline skimage/filters/thresholding.py]: that are less

File: skimage/filters/thresholding.py [hunk 2]
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+
+    Examples
+    --------
+    >>> from skimage.data import camera
+    >>> image = camera()
+    >>> thresh = threshold_li(image)
+    >>> binary = image <= thresh
+    """
+
+    tolerance=0.5
```
Comment (blink1073) [inline skimage/filters/thresholding.py]: pep8: add spaces between the `=`

File: skimage/filters/thresholding.py [hunk 3]
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+    # Calculate the mean gray-level 
+    mean = image.mean()
+    
+    # Initial estimate 
+    new_thresh = mean
+    old_thresh = new_thresh + 2*tolerance
```
Comment (blink1073) [inline skimage/filters/thresholding.py]: add spaces between `*`

File: skimage/filters/thresholding.py [hunk 4]
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+        threshold = int(old_thresh + 0.5)   # range 
+        # Calculate the means of background and object pixels 
+        # Background 
+        mean_back = image[ image <= threshold ].mean()
+        # Object 
+        mean_obj = image[ image > threshold ].mean()
+
+        temp = (mean_back - mean_obj) / (np.log(mean_back) - np.log(mean_obj))
+
+        if temp < 0: # (temp < -2.220446049250313E-16)
```
Comment (blink1073) [inline skimage/filters/thresholding.py]: Why is the commented value there?  Was that in the ImageJ version and no longer needed?

**Comment by blink1073** [comment]
@jmetz, thanks for the contribution!  That error looks like a temporary connectivity issue.  
Can you please add a test for this in [test_thresholding](https://github.com/scikit-image/scikit-image/blob/master/skimage/filters/tests/test_thresholding.py) based on one of the other tests?

File: skimage/filters/thresholding.py (hunk 1 above)
Comment (jmetz) [inline skimage/filters/thresholding.py]: Copied the description from the IJ code

File: skimage/filters/thresholding.py (hunk 4 above)
Comment (jmetz) [inline skimage/filters/thresholding.py]: Thx for the feedback @blink1073  - will fix the PEP8 issues asap. 

See above about the IJ description (re: "that are less" ).

The value is from the IJ version - left it in to get confirmation that we def. don't need to handle floating point issues explicitly..?

File: skimage/filters/thresholding.py [hunk 5]
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+    # Calculate the mean gray-level 
+    mean = image.mean()
+    
+    # Initial estimate 
+    new_thresh = mean
+    old_thresh = new_thresh + 2*tolerance
+
+    # Stop the iterations when the difference between the
+    # new and old threshold values is less than the tolerance 
+    while abs( new_thresh - old_thresh ) > tolerance:
```
Comment (blink1073) [inline skimage/filters/thresholding.py]: Pep8: please remove all spaces between `(` and `[` (also below).

File: skimage/filters/thresholding.py (hunk 4 above)
Comment (blink1073) [inline skimage/filters/thresholding.py]: We typically don't test against an `epsilon`, so I'd say remove the comment.

Also, can you run a "remove trailing whitespace" in your editor?

**Comment by stefanv** [comment]
If this is a line-by-line translation, we should include their copyright in the local directory's README.txt.

**Comment by jmetz** [comment]
@stefanv : It's quite a heavy adaptation (runs on image instead of histogram, replaced large portions original with one-line `numpy` equivalents) - guess it's a bit grey as I used the IJ code as a "pseudo-code" template.

File: skimage/filters/thresholding.py (hunk 5 above)
Comment (jmetz) [inline skimage/filters/thresholding.py]: @blink1073 : How are the tests created in `test_thresholding`? 
E.g. the `test_otsu_coins_image`
    
    def test_otsu_lena_image():
        img = skimage.img_as_ubyte(data.lena())
        assert 140 < threshold_otsu(img) < 142

-- was the range (140--142) determined using an external implementation?

Comment (blink1073) [inline skimage/filters/thresholding.py]: @jmetz, typically we have run the algorithms against one of our test images, verified that the image is visually correct and then used the resulting array as a regression test.

Comment (jmetz) [inline skimage/filters/thresholding.py]: @blink1073 - thanks for the clarification - I'll run the algorithm on some of the test images and add in the tests over the next few days.

File: skimage/filters/thresholding.py [hunk 6]
```diff
@@ -302,3 +303,68 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+    # Stop the iterations when the difference between the
+    # new and old threshold values is less than the tolerance 
+    while abs( new_thresh - old_thresh ) > tolerance: 
+        old_thresh = new_thresh
+        threshold = int(old_thresh + 0.5)   # range 
+        # Calculate the means of background and object pixels 
+        # Background 
+        mean_back = image[ image <= threshold ].mean()
+        # Object 
+        mean_obj = image[ image > threshold ].mean()
```
Comment (ahojnnes) [inline skimage/filters/thresholding.py]: Remove spaces around []

**Comment by jmetz** [comment]
@blink1073 : Added tests (using visual checks for consistency) - also added licence file as suggested by @stefanv. Updated algorithm to subtract  (then finally re-add) `min` to handle negative images.

File: skimage/filters/tests/test_thresholding.py [hunk 7]
```diff
@@ -152,6 +164,29 @@ def test_otsu_astro_image():
...
+    coins = skimage.img_as_ubyte(data.coins())
+    assert 95 < threshold_li(coins) < 97
+
+
+def test_li_coins_image_as_float():
+    coins = skimage.img_as_float(data.coins())
+    assert 0.37 < threshold_li(coins) < 0.38
+
+
+def test_li_lena_image():
```
Comment (blink1073) [inline skimage/filters/tests/test_thresholding.py]: Please remove the lena example, we're slowing removing lena from scikit-image (due to the provocative nature of the image).

File: skimage/filters/thresholding.py (hunk 1 above)
Comment (blink1073) [inline skimage/filters/thresholding.py]: I'd still argue for changing it.  :smile:

File: skimage/filters/thresholding.py [hunk 8]
```diff
@@ -302,3 +303,74 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+    threshold : float
+        Upper threshold value. All pixels intensities that less or equal of
+        this value assumed as foreground.
+
+    References
+    ----------
+    .. [1] Li C.H. and Lee C.K. (1993) "Minimum Cross Entropy Thresholding"
+           Pattern Recognition, 26(4): 617-625
+    .. [2] Li C.H. and Tam P.K.S. (1998) "An Iterative Algorithm for Minimum
+           Cross Entropy Thresholding"Pattern Recognition Letters, 18(8): 771-776
```
Comment (blink1073) [inline skimage/filters/thresholding.py]: Missing space after quotation mark.

File: skimage/filters/thresholding.py [hunk 9]
```diff
@@ -302,3 +303,74 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+    Adapted for skimage by J. Metz from ImageJ plugin by G.Landini
+
+    Examples
+    --------
+    >>> from skimage.data import camera
+    >>> image = camera()
+    >>> thresh = threshold_li(image)
+    >>> binary = image <= thresh
+    """
+    # Requires positive image ( log(mean))
```
Comment (blink1073) [inline skimage/filters/thresholding.py]: remove space after `(`

File: skimage/filters/thresholding.py [hunk 10]
```diff
@@ -302,3 +303,74 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+    offset = image.min()
+    # Can't use fixed tolerance for float image
```
Comment (blink1073) [inline skimage/filters/thresholding.py]: Please use `can not`.

File: skimage/filters/thresholding.py [hunk 11]
```diff
@@ -302,3 +303,74 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+    new_thresh = mean
+    old_thresh = new_thresh + 2 * tolerance
+
+    # Stop the iterations when the difference between the
+    # new and old threshold values is less than the tolerance
+    while abs(new_thresh - old_thresh) > tolerance:
+        old_thresh = new_thresh
+        threshold = old_thresh + tolerance   # range
+        # Calculate the means of background and object pixels
+        # Background
```
Comment (blink1073) [inline skimage/filters/thresholding.py]: I think this and the next comment are just adding confusion.  With the previous comment and the names of the variables, I think you are covered.

File: skimage/filters/thresholding.py [hunk 12]
```diff
@@ -302,3 +303,74 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+    --------
+    >>> from skimage.data import camera
+    >>> image = camera()
+    >>> thresh = threshold_li(image)
+    >>> binary = image <= thresh
+    """
+    # Requires positive image ( log(mean))
+    offset = image.min()
+    # Can't use fixed tolerance for float image
+    imrange = image.max()-offset
```
Comment (blink1073) [inline skimage/filters/thresholding.py]: Please add spaces around `-` operator.

**Comment by blink1073** [comment]
LGTM, thanks @jmetz!

File: skimage/filters/thresholding.py (hunk 1 above)
Comment (stefanv) [inline skimage/filters/thresholding.py]: Err, please fix this.

File: skimage/filters/thresholding.py [hunk 13]
```diff
@@ -302,3 +303,71 @@ def threshold_isodata(image, nbins=256, return_all=False):
...
+           Pattern Recognition, 26(4): 617-625
+    .. [2] Li C.H. and Tam P.K.S. (1998) "An Iterative Algorithm for Minimum
+           Cross Entropy Thresholding" Pattern Recognition Letters, 18(8): 771-776
+    .. [3] Sezgin M. and Sankur B. (2004) "Survey over Image Thresholding
+           Techniques and Quantitative Performance Evaluation" Journal of
+           Electronic Imaging, 13(1): 146-165
+           http://citeseer.ist.psu.edu/sezgin04survey.html
+    .. [4] ImageJ AutoThresholder code, http://fiji.sc/wiki/index.php/Auto_Threshold
+
+    Adapted for skimage by J. Metz from ImageJ plugin by G.Landini
```
Comment (stefanv) [inline skimage/filters/thresholding.py]: This should be in CONTRIBUTORS.txt

Comment (jmetz) [inline skimage/filters/thresholding.py]: @stefanv : I just moved the contribution statement to CONTRIBUTORS.txt and updated the docstring - should I create a new pull request or should this one be reopened?

Comment (blink1073) [inline skimage/filters/thresholding.py]: @jmetz, please open a new PR.

Comment (jmetz) [inline skimage/filters/thresholding.py]: Updates now in #1392

**Pull request merged**